### PR TITLE
Issue #SB-3715 feat: content meta filter with test cases

### PR DESCRIPTION
**Pull request by revathitarento**

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -165,26 +173,65 @@ const telemetryConfig = {
...
+  var blackListedContenttype = blacklistedContentTypeList ? blacklistedContentTypeList.split(',') : []
+  var allowedResourcetype = whitelistedResourceTypeList ? whitelistedResourceTypeList.split(',') : []
+  var blackListedResourcetype = blacklistedResourceTypeList ? blacklistedResourceTypeList.split(',') : []
+
+  var channelConf = generateConfigString(allowedChannels, blackListedChannels)
+  var frameworkConf = generateConfigString(allowedFramework, blackListedFramework)
+  var mimeTypeConf = generateConfigString(allowedMimetype, blackListedMimetype)
+  var contentTypeConf = generateConfigString(allowedContenttype, blackListedContenttype)
+  var resourceTypeConf = generateConfigString(allowedResourcetype, blackListedResourcetype)
+
```
Comment (rajeevsathish) [inline src/app.js]: Refactor the code and push it in a new file. app.js should call that file

File: src/middlewares/filter.middleware.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
+var _ = require('underscore')
+var filterService = require('../service/filterService')
+var LOG = require('sb_logger_util')
+var utilsService = require('../service/utilsService')
+var path = require('path')
+var filename = path.basename(__filename)
+
+function addMetaFilters (req, res, next) {
+  // If the request body has filter by metaFilter data, continue with the same filter, do not alter the values
+  // else call the getMetaSearchJSON() function to generate the search string for the meta filters
```
Comment (rajeevsathish) [inline src/middlewares/filter.middleware.js]: check the comment

File: src/app.js (hunk 1 above)
Comment (revathitarento) [reply src/app.js]: Done, removed the lines and created new file

File: src/middlewares/filter.middleware.js (hunk 2 above)
Comment (revathitarento) [reply src/middlewares/filter.middleware.js]: updated the comment

File: src/app.js [hunk 3]
```diff
@@ -8,11 +8,9 @@ var TelemetryUtil = require('sb_telemetry_util')
 var telemetry = new TelemetryUtil()
 var fs = require('fs')
 var configUtil = require('sb-config-util')
-var _ = require('underscore')
-var filename = path.basename(__filename)
-var utilsService = require('./service/utilsService')
-var LOG = require('sb_logger_util')
+
 const contentProvider = require('sb_content_provider_util')
+var contentMeta = require('./contentMetaFilter')
```
Comment (rajeevsathish) [inline src/app.js]: why is the name contentMeta what is the use of the file please specify that

File: src/middlewares/filter.middleware.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    if (err) {
+      LOG.error(utilsService.getLoggerData({}, 'ERROR', filename, 'addChannelFilters',
+        'failed to get channels'))
+    } else if (searchJSON && (!_.isEmpty(searchJSON))) {
+      for (var key in searchJSON) {
+        if (searchJSON.hasOwnProperty(key)) {
+          var val = searchJSON[key]
+
+          if (key === filter && val !== undefined) {
+            var finalval = req.body.request.filters[key] = val
```
Comment (rajeevsathish) [inline src/middlewares/filter.middleware.js]: can you please explain the logic

File: src/test/appSpec.js [hunk 5]
```diff
@@ -1,14 +1,76 @@
...
-
+var fs = require('fs')
 var request = require('request')
 var host = 'http://localhost:5000'
 const whiteListedChannelList = process.env.sunbird_content_service_whitelisted_channels
 const blackListedChannelList = process.env.sunbird_content_service_blacklisted_channels
-var whiteListQuery = whiteListedChannelList ? whiteListedChannelList.split(',') : []
-var blackListedChannelListNew = blackListedChannelList ? blackListedChannelList.split(',') : []
-var blackListQuery = {'ne': (blackListedChannelListNew)}
+
```
Comment (rajeevsathish) [inline src/test/appSpec.js]: this Spec file has to be refactored with the changes done please fix this and please don't comment any old specs

File: src/contentMetaFilter.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+function getMetaFilterConfig () {
+  LOG.info(utilsService.getLoggerData({}, 'INFO',
+    filename, 'getFilterConfig', 'environment info', process.env))
+  var allowedChannels = process.env.sunbird_content_service_whitelisted_channels
+    ? process.env.sunbird_content_service_whitelisted_channels.split(',') : []
+  var blackListedChannels = process.env.sunbird_content_service_blacklisted_channels
+    ? process.env.sunbird_content_service_blacklisted_channels.split(',') : []
+  var allowedFramework = process.env.sunbird_content_filter_framework_whitelist
+    ? process.env.sunbird_content_filter_framework_whitelist.split(',') : []
+  var blackListedFramework = process.env.sunbird_content_filter_framework_blacklist
```
[the remark this answers is not in the document]
Comment (revathitarento) [reply src/contentMetaFilter.js]: Done

File: src/app.js (hunk 3 above)
Comment (revathitarento) [reply src/app.js]: Changed to contentMetaProvider

File: src/middlewares/filter.middleware.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+function fetchFilterQuery (req, filter) {
+  filterService.getMetadataFilterQuery(function (err, searchJSON) {
+    // console.log('err', err)
+    if (err) {
+      LOG.error(utilsService.getLoggerData({}, 'ERROR', filename, 'fetchFilterQuery',
+        'failed to get fetch filter query'))
+    } else if (searchJSON) {
+      for (var key in searchJSON) {
```
Comment (rajeevsathish) [inline src/middlewares/filter.middleware.js]: I am not sure what is the use of this code and the If statement 
const object1 = new Object();
object1.property1 = 42;
for(var key in object1){
console.log(key)
console.log(object1.hasOwnProperty(key));
}
the if statement is always true

File: src/test/contentMetaConfig.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+const whitelistedFrameworkList = process.env.sunbird_content_service_whitelisted_framework
+const blacklistedFrameworkList = process.env.sunbird_content_service_blacklisted_framework
+const whitelistedMimeTypeList = process.env.sunbird_content_service_whitelisted_mimetype
+const blacklistedMimeTypeList = process.env.sunbird_content_service_blacklisted_mimetype
+const whitelistedContentTypeList = process.env.sunbird_content_service_whitelisted_contenttype
+const blacklistedContentTypeList = process.env.sunbird_content_service_blacklisted_contenttype
+const whitelistedResourceTypeList = process.env.sunbird_content_service_whitelisted_resourcetype
+const blacklistedResourceTypeList = process.env.sunbird_content_service_blacklisted_resourcetype
+
+module.exports = Object.freeze({
```
Comment (rajeevsathish) [inline src/test/contentMetaConfig.js]: why is the object name Freeze?

File: src/test/contentMetaFilterSpec.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
+var configUtil = require('../libs/sb-config-util')
+var _ = require('lodash')
+
+var contentMetaConfig = require('./contentMetaConfig')
+
+var allowedChannels = contentMetaConfig.allowedChannels
```
Comment (rajeevsathish) [inline src/test/contentMetaFilterSpec.js]: why do we need to create separate variables here can't we use directly the value from the constant

File: src/test/metaFilterRouteSpec.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+]
+
+var filterMiddleware = require('../middlewares/filter.middleware')
+var httpMocks = require('node-mocks-http')
+
+var baseUrl = 'http://localhost:5000/v1/'
+var async = require('async')
+var contentMetaConfig = require('./contentMetaConfig')
+
+var allowedChannels = contentMetaConfig.allowedChannels
```
Comment (rajeevsathish) [inline src/test/metaFilterRouteSpec.js]: same comment why again a variable do all the data manipulation in the Data file and use it directly...

File: src/test/contentMetaConfig.js (hunk 8 above)
Comment (revathitarento) [reply src/test/contentMetaConfig.js]: its a js function

Comment (rajeevsathish) [reply src/test/contentMetaConfig.js]: ok, the freeze function is used I got it, but why only doing this manipulate the string and use it directly.
you are creating the var again in the spec which does not solve the purpose of creating this file please revisit this code

File: src/test/contentMetaFilterSpec.js (hunk 9 above)
Comment (revathitarento) [reply src/test/contentMetaFilterSpec.js]: i use allowedChannels in more than 10 places, so its better to assign it to a value

File: src/middlewares/filter.middleware.js (hunk 7 above)
Comment (revathitarento) [reply src/middlewares/filter.middleware.js]: yes, removed

File: src/contentMetaFilter.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    ? process.env.sunbird_content_service_whitelisted_contenttype.split(',') : []
+  var blackListedContenttype = process.env.sunbird_content_service_blacklisted_contenttype
+    ? process.env.sunbird_content_service_blacklisted_contenttype.split(',') : []
+  var allowedResourcetype = process.env.sunbird_content_service_whitelisted_resourcetype
+    ? process.env.sunbird_content_service_whitelisted_resourcetype.split(',') : []
+  var blackListedResourcetype = process.env.sunbird_content_service_blacklisted_resourcetype
+    ? process.env.sunbird_content_service_blacklisted_resourcetype.split(',') : []
+
+  // Check if the Filter Config service data is defined, if yes, create Object with it
+  const filterConfigService = ''
```
Comment (rajeevsathish) [inline src/contentMetaFilter.js]: TODO please don't make an empty string and use the empty string check,
make the filterConfigService as null and do the null check

Comment (kochhar) [reply src/contentMetaFilter.js]: Is this comment addressed?

**Pull request merged**